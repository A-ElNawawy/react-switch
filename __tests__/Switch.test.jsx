import React from 'react';
import { shallow, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Switch from '../src';

const noop = () => {};

describe('Props', () => {
  it('matches snapshot with default props', () => {
    const wrapper = render(<Switch checked={false} onChange={noop} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('matches snapshot with custom props', () => {
    const wrapper = render(
      <Switch
        checked
        onChange={noop}
        disabled
        onColor="#abc"
        offHandleColor="#def"
        handleDiameter={9}
        checkedIcon={(
          <div style={{ color: 'pink', paddingTop: 2, paddingLeft: 6, fontSize: 20 }}>
            ‚ù§
          </div>
        )}
        boxShadow="0px 1px 5px rgba(0, 0, 0, 0.6)"
        height={13}
        width={37}
        className="qux"
        id="foo"
        aria-labelledby="baz"
        aria-label="bar"
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('onClick', () => {
  it('calls onChange when bg is clicked with !checked as argument', () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(<Switch onChange={mockOnChange} checked={false} />);
    wrapper.find('.react-switch-bg').simulate('click');
    expect(mockOnChange).toBeCalledWith(true);
  });

  it('does not call onChange when disabled', () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(<Switch onChange={mockOnChange} checked={false} disabled />);
    wrapper.find('.react-switch-bg').simulate('click');
    expect(mockOnChange).not.toBeCalled();
  });
});

describe('handle', () => {
  it('sets a boxShadow on focus and loses it on blur', () => {
    const wrapper = shallow(<Switch onChange={noop} checked={false} />);
    wrapper.find('.react-switch-handle').simulate('focus');
    expect(wrapper.find('.react-switch-handle').get(0).props.style.boxShadow).toBe('0px 0px 2px 3px #33bbff');
    wrapper.find('.react-switch-handle').simulate('blur');
    expect(wrapper.find('.react-switch-handle').get(0).props.style.boxShadow).toBeNull();
  });

  it('calls onChange and prevent default on spacebar and enter', () => {
    const mockOnChange = jest.fn();
    const mockPreventDefault = jest.fn();
    const wrapper = shallow(<Switch onChange={mockOnChange} checked={false} />);
    wrapper.find('.react-switch-handle').simulate('focus');
    wrapper.find('.react-switch-handle').simulate('keydown', { keyCode: 32, preventDefault: mockPreventDefault });
    wrapper.find('.react-switch-handle').simulate('keydown', { keyCode: 13, preventDefault: mockPreventDefault });
    expect(mockPreventDefault.mock.calls.length).toBe(2);
    expect(mockOnChange.mock.calls.length).toBe(2);
  });
  it('doesn\'t call onChange on some other key', () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(<Switch onChange={mockOnChange} checked={false} />);
    wrapper.find('.react-switch-handle').simulate('focus');
    wrapper.find('.react-switch-handle').simulate('keydown', { keyCode: 14, preventDefault: noop });
    expect(mockOnChange).not.toBeCalled();
  });
});

describe('checked prop', () => {
  it('affects left-position when it changes', () => {
    const wrapper = shallow(<Switch onChange={noop} checked={false} />);
    wrapper.setProps({ checked: true });
    expect(wrapper.find('.react-switch-handle').get(0).props.style.left).toBe(29);
    wrapper.setProps({ checked: false });
    expect(wrapper.find('.react-switch-handle').get(0).props.style.left).toBe(1);
  });

  it('affects background when it changes', () => {
    const wrapper = shallow(
      <Switch onChange={noop} checked={false} offColor="#aabbcc" onColor="#ddeeff" />
    );
    wrapper.setProps({ checked: true });
    expect(wrapper.find('.react-switch-bg').get(0).props.style.background).toBe('#ddeeff');
    wrapper.setProps({ checked: false });
    expect(wrapper.find('.react-switch-bg').get(0).props.style.background).toBe('#aabbcc');
  });
});

// Test dragging behaviour
describe('drag related behaviour', () => {
  it('triggers onChange callback when isDragging is never set to true', () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(<Switch onChange={mockOnChange} checked={false} />);
    wrapper.find('.react-switch-handle').simulate('touchstart', { touches: [{ clientX: 100 }] });
    wrapper.find('.react-switch-handle').simulate('touchend', { preventDefault: noop });
    expect(mockOnChange).toBeCalledWith(true);
  });
  it('triggers onChange when handle is dragged half way or more', () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(
      <Switch
        onChange={mockOnChange}
        checked={false}
        height={15}
        width={30}
      />
    );
    wrapper.find('.react-switch-handle').simulate('touchstart', { touches: [{ clientX: 100 }] });
    wrapper.find('.react-switch-handle').simulate('touchmove', { touches: [{ clientX: 107.5 }] });
    wrapper.find('.react-switch-handle').simulate('touchend', { preventDefault: noop });
    expect(mockOnChange).toBeCalledWith(true);
  });
  it('doesn\'t trigger onChange when handle is dragged less than half way', () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(
      <Switch
        onChange={mockOnChange}
        checked={false}
        height={15}
        width={30}
      />
    );
    wrapper.find('.react-switch-handle').simulate('touchstart', { touches: [{ clientX: 100 }] });
    wrapper.find('.react-switch-handle').simulate('touchmove', { touches: [{ clientX: 107.4 }] });
    wrapper.find('.react-switch-handle').simulate('touchend', { preventDefault: noop });
    expect(mockOnChange).not.toBeCalled();
  });
});
