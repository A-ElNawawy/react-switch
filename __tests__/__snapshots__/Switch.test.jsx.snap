// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch matches snapshot with custom props excl disabled 1`] = `
<div
  class="qux"
  style="height:13px;width:37px;background:grey;border-radius:6.5px;display:inline-block;position:relative;opacity:1;-webkit-transition:all 0.2s;-moz-transition:all 0.2s;transition:all 0.2s;cursor:pointer"
>
  <div
    class="react-switch-fg"
    style="height:13px;width:37px;opacity:1;background:#abc;-webkit-transition:opacity 0.2s ease-out;-moz-transition:opacity 0.2s ease-out;transition:opacity 0.2s ease-out;border-radius:6.5px"
  />
  <div
    aria-checked="true"
    aria-disabled="false"
    aria-label="bar"
    aria-labelledby="baz"
    class="react-switch-handle"
    id="foo"
    role="checkbox"
    style="touch-action:none;height:11px;width:11px;background:#def;-webkit-transition:left 0.2s ease-out;-moz-transition:left 0.2s ease-out;transition:left 0.2s ease-out;display:inline-block;border-radius:50%;position:absolute;left:25px;top:1px;border:0;outline:0"
    tabindex="0"
  />
</div>
`;

exports[`Switch matches snapshot with default props 1`] = `
<div
  style="height:28px;width:56px;background:grey;border-radius:14px;display:inline-block;position:relative;opacity:1;-webkit-transition:all 0.2s;-moz-transition:all 0.2s;transition:all 0.2s;cursor:pointer"
>
  <div
    class="react-switch-fg"
    style="height:28px;width:56px;opacity:0;background:green;-webkit-transition:opacity 0.2s ease-out;-moz-transition:opacity 0.2s ease-out;transition:opacity 0.2s ease-out;border-radius:14px"
  />
  <div
    aria-checked="false"
    aria-disabled="false"
    class="react-switch-handle"
    role="checkbox"
    style="touch-action:none;height:26px;width:26px;background:white;-webkit-transition:left 0.2s ease-out;-moz-transition:left 0.2s ease-out;transition:left 0.2s ease-out;display:inline-block;border-radius:50%;position:absolute;left:1px;top:1px;border:0;outline:0"
    tabindex="0"
  />
</div>
`;
